resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/RefillRateLimiter.java

```diff
@@ -0,0 +1,476 @@
+        if(nanosPerPermission==0) {
+            return Long.MAX_VALUE;
+        } else if(nanosSinceLastUpdate<0l) {
+            return 0l;
+        }
+
+        return nanosSinceLastUpdate / nanosPerPermission;
+    }
+
+    /**
+     * Calculates time to wait for the required permits of permissions to get accumulated
+     *
+     * @param permits              permits of required permissions
+     * @param nanosPerPermission   nanos needed for one permission to be released
+     * @param availablePermissions currently available permissions, can be negative if some
+     *                             permissions have been reserved
+     */
+    private long nanosToWaitForPermission(final int permits, final long nanosPerPermission,
+                                          final long availablePermissions) {
+        if (availablePermissions >= permits) {
+            return 0L;
+        }
+
+        long permissionsWanted = permits - availablePermissions;
+        long nanosToWait = permissionsWanted*nanosPerPermission;
+        return nanosToWait;
+    }
+
+    /**
+     * Determines whether caller can acquire permission before timeout or not and then creates
+     * corresponding {@link State}. Reserves permissions only if caller can successfully wait for
+     * permission.
+     *
+     * @param state
+     * @param permits        permits of permissions
+     * @param timeoutInNanos max time that caller can wait for permission in nanoseconds
+     * @param permissions    permissions for new {@link State}
+     * @param nanosToWait    nanoseconds to wait for the next permission
+     * @return new {@link State} with possibly reserved permissions and time to wait
+     */
+    private State reservePermissions(final State state, final int permits,
+                                     final long timeoutInNanos,
+                                     final int permissions, final long nanosToWait) {
+        boolean canAcquireInTime = timeoutInNanos >= nanosToWait;
+        int permissionsWithReservation = permissions;
+        if (canAcquireInTime) {
+            permissionsWithReservation -= permits;
+        }
+
+        return new State(state.config, permissionsWithReservation, nanosToWait, currentNanoTime());
+    }
+
+
+    /**
+     * If nanosToWait is bigger than 0 it tries to park {@link Thread} for nanosToWait but not
+     * longer then timeoutInNanos.
+     *
+     * @param timeoutInNanos max time that caller can wait
+     * @param nanosToWait    nanoseconds caller need to wait
+     * @return true if caller was able to wait for nanosToWait without {@link Thread#interrupt} and
+     * not exceed timeout
+     */
+    private boolean waitForPermissionIfNecessary(final long timeoutInNanos,
+                                                 final long nanosToWait) {
+        boolean canAcquireImmediately = nanosToWait <= 0;
+        boolean canAcquireInTime = timeoutInNanos >= nanosToWait;
+
+        if (canAcquireImmediately) {
```

Comment:
Run this if right after `canAcquireImmediately`. It is worth to change it in `AtomicRateLimiter` as well.